sdks/go/pkg/beam/io/synthetic/source_test.go

```diff
@@ -162,6 +163,59 @@ func TestSourceConfig_BuildFromJSON(t *testing.T) {
 	}
 }
 
+// TestSourceConfig_NumHotKeys tests that setting the number of hot keys
+// for a synthetic source works correctly.
+func TestSourceConfigBuilder_NumHotKeys(t *testing.T) {
+	tests := []struct {
+		elms    int
+		hotKeys int
+	}{
+		{elms: 15, hotKeys: 2},
+		{elms: 30, hotKeys: 10},
+		{elms: 50, hotKeys: 25},
+	}
+	for _, test := range tests {
+		test := test
+		t.Run(fmt.Sprintf("(elm = %v)", test.hotKeys), func(t *testing.T) {
+			dfn := sourceFn{}
+			cfg := DefaultSourceConfig()
+			cfg.NumElements(test.elms)
+			cfg.HotKeyFraction(1.0)
+			cfg.NumHotKeys(test.hotKeys)
+
+			keys, _, err := simulateSourceFn(t, &dfn, cfg.Build())
+			if err != nil {
+				t.Errorf("Failure processing sourceFn: %v", err)
+			}
+
+			m := make(map[string]int)
+			for _, key := range keys {
+				encoded := hex.EncodeToString(key)
+				m[encoded]++
+			}
+
+			numOfHotKeys := 0
+			numOfAllKeys := 0
+			for _, element := range m {
+				numOfAllKeys += element
+				if element > 1 {
+					numOfHotKeys += 1
+				}
+			}
+
+			if numOfAllKeys != test.elms {
```

Comment:
I don't think the numOfAllKeys check here is necessary. It essentially boils down to counting the number of elements emitted which is already done by other unit tests. The only reason I can see for this is if you suspect that enabling hot keys might cause elements to drop, when they wouldn't otherwise, but based on your implementation I don't think that's something to worry about.

Reply:
Thank you, I removed this check.